website/docs/wiki/caching.md

```diff
@@ -0,0 +1,84 @@
+---
+title: Caching
+---
+
+Either by lack of existing solutions for gRPC caching or by need for a non-trivial logic you might require to use an application cache for your responses.
+
+## Basic middleware cache
+
+Without any extra support, with a bit of effort you can implement middleware level caching just with existing middleware interface and its `next` function, allowing you to skip the following handlers.
+
+```typescript
+app.use(async (call, next) => {
+  const key = getCacheKey(call)
+  const cachedResponse = await cacheGet(key)
+  if (cachedResponse) {
+    call.response = cachedToProto(cachedResponse)
+    // skip following handlers, we have response from cache
+    return
+  }
+  // cache miss, execute handlers
+  await next()
+  // save result for subsequent requests
+  setCacheKey(protoToCached(call.response), TTL)
+})
+```
+
+This simple middleware is a good start: we effectively skip the handlers, can manage TTL and can combine with any custom logic for the matching. For example we can hook the cache middleware after authentication middleware and create per-user cache with prepared data about user from the call.
+
+Assume we have non-in memory cache and we need to serialize the instances into some form and back. That is why there's `cachedToProto` and `protoToCached` functions. This is the troublesome part: doing this serialization generically might be challenging (and could be potentially solvable only via custom solution for each RPC/message) and it might not be "cheap" (cache hits should be blistering fast and not perform double encoding `cache-to-proto` by you and `proto-to-wire` by gRPC).
+
+That's why ProtoCat has a mechanism to remove both of the problems: generic approach for all RPCs without mapping functions and avoid double encoding.
+
+## Cache middleware
+
+If you don't care too much about the details, you can just use
+
+```typescript
+import { createCache, CacheImplementation } from 'protocat'
+
+const cache: CacheImplementation = {
+  set: (key, value, call) => {
+    await someStorage.set(key, value) // MUST SET BUFFER
+  },
+  get: (key, call) => someStorage.get(key), // MUST RETURN BUFFER
+  hash: call =>
+    `${call.path}::${call.user?.id ?? ''}::${call.request.toArray().join('-')}`,
```

Comment:
Did not think about it, when I wrote it, I just wanted to make something that could not be confused with the RPC path. Same syntax is used for C++ namespaces and PHP static members I think :cat: 

Not super related, but I think it is fine, since it's just a demo. Or do you think we should spend more brains on this, before everyone copies that into their gRPC projects :smirk_cat: